File: 1011-capacity-to-ship-packages-within-d-days/1011-capacity-to-ship-packages-within-d-days.java
class Solution {
    private boolean ispossible(int[] weights,int mid,int days)
    {
        int day=1;
        int ship_load=0;
        for(int i=0;i<weights.length;)
        {
            if(ship_load+weights[i]<=mid)
            {
                ship_load+=weights[i];
                i++;
            }
            else
            {
                day++;
                ship_load=0;
            }
            if(day>days)
            {
                return false;
            }
        }
        return true;
    }
    public int shipWithinDays(int[] weights, int days) {
        int ans=0;
        int total=0;
        for(int i=0;i<weights.length;i++)
        {
            total+=weights[i];
        }
        int s=1;
        int e=total;
        while(s<=e)
        {
            int mid=(s+e)/2;
            if(ispossible(weights,mid,days))
            {
                ans=mid;
                e=mid-1;
            }
            else
            {
                s=mid+1;
            }
        }
        return ans;
    }
}

File: 1011-capacity-to-ship-packages-within-d-days/1011-capacity-to-ship-packages-within-d-days.py
class Solution:
    def ispossible(self,weights:List[int],mid:int,days:int)->bool:
        day=1
        ship_load=0
        for i in range(len(weights)):
            if(ship_load+weights[i]<=mid):
                ship_load+=weights[i]
                i+=1
            else:
                day+=1
                ship_load=weights[i]
            if(day>days):
                return False           
        return True

    def shipWithinDays(self, weights: List[int], days: int) -> int:
        ans=0
        total=sum(weights)
        s=max(weights)
        e=total
        while(s<=e):
            mid=(s+e)//2
            if(self.ispossible(weights,mid,days)):
                ans=mid
                e=mid-1
            else:
                s=mid+1
        return ans